src/io/fastx.rs

```diff
@@ -0,0 +1,319 @@
+//! format.
+//!
+//! ```
+//! use bio::io::{fasta, fastq, fastx};
+//! use std::io;
+//!
+//! fn count_bases<T, E, I>(mut records: I) -> Result<usize, E>
+//! where T: fastx::Record,
+//!     E: std::error::Error,
+//!     I: Iterator<Item = Result<T, E>> {
+//!     let mut nb_bases = 0;
+//!     for result in records {
+//!         let record = result?;
+//!         nb_bases += record.seq().len();
+//!     }
+//!     Ok(nb_bases)
+//! }
+//!
+//! let mut raw_reader = io::Cursor::new(b">id desc
+//! ACTG
+//! ");
+//!
+//! match fastx::get_type(&mut raw_reader) {
+//!     Ok(fastx::FastxType::Fasta) => {
+//!         let mut reader = fasta::Reader::new(raw_reader);
+//!         let nb_bases = count_bases(reader.records()).unwrap();
+//!         println!("Number of bases: {}", nb_bases);
+//!     },
+//!     Ok(fastx::FastxType::Fastq) => {
+//!         let mut reader = fastq::Reader::new(raw_reader);
+//!         let nb_bases = count_bases(reader.records()).unwrap();
+//!         println!("Number of bases: {}", nb_bases);
+//!     },
+//!     _ => println!("Encountered an error"),
+//! }
+//! ```
+//!
+//! ## Filtration
+//!
+//! In this example, we define an at_least_n_bases function that can filter fasta or fastq
+//! records based on their sequence lengths. It works seemlessly with fasta records as if
+//! it was implemented just on fasta records. In a realistic scenario this function might be
+//! defined in a library so callers could use it with both fasta and fastq files as needed.
+//!
+//! ```
+//! use bio::io::{fasta, fastq, fastx};
+//! use std::io;
+//!
+//! fn at_least_n_bases<T, E, I>(mut records: I, n: usize) -> impl Iterator<Item = Result<T, E>>
+//! where T: fastx::Record,
+//!     E: std::error::Error,
+//!     I: Iterator<Item = Result<T, E>> {
+//!     records.filter(move |rr| match rr {
+//!         Ok(r) => r.seq().len() > n,
+//!         _ => true,
+//!     })
+//! }
+//!
+//! let mut reader = fasta::Reader::new(io::stdin());
+//! let mut writer = fasta::Writer::new(io::stdout());
+//!
+//! for record in at_least_n_bases(reader.records(), 10) {
+//!     writer.write_record(&record.unwrap());
+//! }
+//! ```
+//!
+use std::io;
+use std::io::SeekFrom;
+use std::io::prelude::*;
+use std::mem;
+
+use crate::io::{fasta, fastq};
+use crate::utils::TextSlice;
+
+macro_rules! passthrough {
+    ($name:ident, $t:ty) => {
+        fn $name(&self) -> $t {
+            self.$name()
+        }
+    };
+}
+
+macro_rules! matchthrough {
+    ($name:ident, $t:ty) => {
+        fn $name(&self) -> $t {
+            match self {
+                EitherRecord::FASTA(f) => f.$name(),
+                EitherRecord::FASTQ(f) => f.$name(),
+            }
+        }
+    };
+}
+
+pub trait Record {
+    fn is_empty(&self) -> bool;
+    fn check(&self) -> Result<(), &str>;
+    fn id(&self) -> &str;
+    fn desc(&self) -> Option<&str>;
+    fn seq(&self) -> TextSlice<'_>;
+}
```

Comment:
In my point of view it could be nice to add in Record trait some method:

- `fn type(&self) -> FastxType`, I imagine a use case where I read file in function *A* and I give `Record` to function *B* it could be nice to directly ask `Record` than pass record type by parameter to *B* function
- `fn qual(&self) -> Option<&[u8]>` If record is fastq just return qual if it's fastq just None
- `fn qual_default(&self, qual_value: u8) -> &[u8]` If record is fastq just return qual else generate a quality string by repeat qual_value

Reply:
Makes sense on the first two on the final one it is a bit challenging because there is nowhere for the default slice to live since for fasta it is a temp value generated by the function. Users can do this:

```Rust
record.qual().unwrap_or(& vec![b'I' ; record.seq().len()])
```

Reply:
I agree, for me if this code is show in documentation it's perfect.